astropy/tests/tests/test_imports.py

```diff
@@ -32,3 +41,37 @@ def test_toplevel_namespace():
     assert "log" in d
     assert "test" in d
     assert "sys" not in d
+
+
+def test_toplevel_lazy_imports():
+    # Check that subpackages are loaded on demand.
+    cmd = dedent("""
+    import astropy, sys
+    assert 'astropy.units' not in sys.modules
+    astropy.units
+    assert 'astropy.units' in sys.modules
+    """)
+    assert subprocess.check_call([sys.executable, "-c", cmd]) == 0
+
+
+def test_toplevel_attribute_error():
+    # Ensure that our __getattr__ does not leak an import error or so.
+    with pytest.raises(AttributeError, match="module 'astropy' has no"):
+        astropy.nonsense
+
+
+def test_completeness_toplevel__all__():
+    # Implicitly check that all items in __all__ exist.
+    all_items = {getattr(astropy, attr) for attr in astropy.__all__}
+    # Verify that the list of modules in __all__ is complete.
+    module_names = {
+        item.__name__.partition(".")[2]
+        for item in all_items
+        if isinstance(item, ModuleType)
+    }
+    module_dirs = {
+        f.name
+        for f in Path(astropy.__file__).parent.glob("[a-z]*")
+        if f.is_dir() and f.name != "extern"
```

Comment:
```suggestion
        if f.is_dir() and f.name != "extern" and f.joinpath("__init__.py").is_file()
```

Reply:
I'd hope we won't add non-python module directories - if we ever do, then maybe that's the right time to add the extra clause... (similarly, in principle one can have modules defined in `*.py` files, which I also ignore; this just tests the existing scheme).

Reply:
The reason I'm worried about this isn't that non-python module dirs are added to vcs, but that it might be done on a development copy and change this test's status possibly in a non-obvious way. I think it's worth spending a minute emaking the test narrow and correct now to save what would likely be a longer time spent on confused debugging for this hypothetical future contributor.

Reply:
This isn't a blocking request though; if I haven't convinced you yet, then we're probably wasting more time discussing it that it'll ever save.

Reply:
I see your point now, though perhaps not quite enough to run all CI again...  Let's just get it in!